### DISCUSSION TITLE
Override CreateWindow to give custom HWND in v11

### DISCUSSION THREAD

**Discussion by trympet** [AvaloniaUI/Avalonia#11686] [General]
I am using GDI to draw a simple shell / spash screen for my app before launching avalonia. I was able to do this in v10 by deriving from `Avalonia.Win32.WindowImpl`, overriding `CreateWindowOverride`, and returning a HWND which I had created.

However, in v11, `Avalonia.Win32.WindowImpl` is internal, so I can no longer override `CreateWindowOverride`. Hence, the question: is there an analogue for supplying my own window handle for v11?

**Comment by timunie** [comment]
honest question: Do you think it's worth the effort it takes? what is the benefit of having such a splash screen? Don't get me wrong, a splash screen makes perfect sense, but you can also use a minimal Avalonia window or your own window based on `Window`.

**Reply by trympet** [reply]
I haven't measured the difference in startup time between v10 and v11, however, resolving resources for the styles, and the subsequent layout./arrange pass, accounted for ~900ms in v10. Meanwhile, the GDI spash screen has a time-to-first-paint of <100ms.

Throw WinComp initialization and D3D11 device creation delay on top, and we are looking at a good 1.5s of time saved in most scenarios.

But I agree with your sentiment -- it's overkill for many scenarios.

**Reply by trympet** [reply]
And the effort is already put in for my part, as it's working fine in v10.

**Reply by trympet** [reply]
Also, to add to the value proposition for my request: it would also make embedding easier.

**Reply by timunie** [reply]
In 11.0 we have done a lot of stuff to improve performance. Can you try in a simple sample if Avalonia Window would be fast enough now? Otherwise I'm afraid you need to roll your own Win32 window or use Reflection to create a WindowImpl.

**Reply by trympet** [reply]
Certainly! That sounds like a decent workaround. I'm almost done with my migration to v11 anyways -- motivated by NativeAOT -- so I'll give it a shot without the GDI shell. I am otherwise happy with the work that has gone into v11. I generally support the decision of making the API surface as small as possible 😄
